### PR TITLE
Edit products: Update product info on table

### PR DESCRIPTION
**Pull request by EdiongAlexander**
## Description
[Please put a description of the changes this branch introduces here.]

## To Test
[Please write instructions for testing your changes locally here. Include any data setup, tests, urls, and potential snags in local testing.]

## Checklist
- [ ] I have looked over the diffs.
- [ ] There are no errors/failing tests from my changes on this branch.
- [ ] I have requested a review on this PR and re-assigned the Jira ticket.

## Screenshots
[Any screenshots of the changes can go here.]

## Extra Notes
[Please include any extra notes you may have about the changes here.]

### CODE REVIEW

**Comment by ObiTam** [comment]
image
the entire header should be clickable and have cursor pointer.

image
please work on the regex. number types shouldn't accept regex. etc. 

Also add errors to the form. 

Submit button should be disabled if 
1  No edit has been made.
2  Errors in the form.

image
add cursor pointer to delete icon.

image
The drag and drop isn't working here.

Confirm from samuel what Save as draft should do. Right now it does nothing. I envisage that it should close the page and create a new draft product with the edited details.

image
Please let's group the products routes together

File: src/components/edit-products/index.js [hunk 1]
```diff
@@ -1,32 +1,56 @@
...
 
 
 
 const EditSingleProduct = () => {
 
   const { sku } = useParams();
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
 
+  const PRODUCT_DATA = useSelector((state) => state.productsData);
```
Comment (ObiTam) [inline src/components/edit-products/index.js:16]: let's use camel case for this please.

File: src/redux/action.js [hunk 2]
```diff
@@ -58,6 +58,32 @@ export const sidebarToggle = (bool) => dispatch => {
   });
 };
 
+export const editProduct = (sku, productInfo, PRODUCT_DATA) => dispatch => {
```
Comment (ObiTam) [inline src/redux/action.js:61]: This should receive the new productInfo and the SKU. 
All the logic should be done in the reducer based on the productData gotten from the previous state. 
This is for encapsulation and abstraction

File: src/components/edit-products/index.js [hunk 3]
```diff
@@ -1,32 +1,56 @@
...
+
+  const handleProductInfo = (key, val) => {
+    setProductInfo((prevProductInfo) => ({
+      ...prevProductInfo,
+      [key]: val,
+    }))
+
   }
+  
+  useEffect(() => {
```
Comment (ObiTam) [inline src/components/edit-products/index.js:44]: lets take this out once we're done.

File: src/components/products-changes/index.js [hunk 4]
```diff
@@ -73,16 +73,20 @@ const ProductChanges = ({ name, productInfo }) => {
     const newCategory = event.target.id;
     setSelectedCategory(newCategory);
```
Comment (ObiTam) [inline src/components/products-changes/index.js:74]: I don't think we need selectedCategory and Selected Files. They are just duplicates of what is coming from product info. I think we can take it out.

**Review by ObiTam — CHANGES_REQUESTED** [review]
Hi Ediong,

The UI looks Amazing. Well done!

I made some functional requests. I also made some requests for optimization of the code. I'll stop here cause its getting a bit lengthy. Please take a look then we can go from there.

Cheers.

File: src/components/products-changes/index.js (hunk 4 above)
Comment (EdiongAlexander) [reply src/components/products-changes/index.js:74]: But the category and files can be updated

File: src/components/edit-products/index.js (hunk 3 above)
Comment (EdiongAlexander) [reply src/components/edit-products/index.js:44]: Okay

**Review by ObiTam — CHANGES_REQUESTED** [review]
image
why is it saying required when there is data. why is the error triggering.

image

It also doesn't let me submit because of the errors.
I'll take a look at the code soon.

**Comment by ObiTam** [comment]
Hi Ediong. 
I'd like us to test this out together.
Before then lets make these changes. 

1. There are branch conflicts you should check and resolve.

2. The product changes component should have a local state which all the input components point to. This state is initialised with the product info prop for edit products. For new products this is initialised with a suitable initial object. Ofcourse you can do that basis your name variable. However, I think this name prop should be changed for readability - perhaps edit product and make it be true or false. Think through this and come up with an optimum solution.

3. We render images using a url. when a new image is uploadedd we do not have a url for that image just the raw image data. we can generate a pseudo url for that image on the fly and upload till the backend can generate that url and send it to us. 
image
This image is from the profile tab. you can view for your use.

let us structure our product with images like this 
```
const PRODUCT_DATA = [
    {
        id: "product-1",
        // other product data
        images: [
            {
                url: 'frontend/image', // pseudo url generated on frontend since we have an imageFile attached
                data: imageFile // image file that will be uploaded on the backend
            },
            {
                url: 'backend/image', // actual image url
                data: null // null since image url is on backend already
            },
            // other images
        ]
    },
    // other products
]
```

4. For the Tabs
image
Let's have a tab state that we start initially as an empty string. If a tab is clicked then we check if tab state is equal to tab clicked. if it is we turn tab to empty string else we turn tab to tab clicked. 

5. Continuing from 2 above. any changes in the product changes component will be made to the local state of the product information. You may have a local state count with change count that is adjusted within the product info state handler that handles changes of the product. This will enable you to change the Submit and Save as Draft to active only if a change has been made. The Product Changes Component performs 3 actions.
-  Cancel will just redirect to the intended new url perhaps the product page. The local state with all changes will be deleted and no changes will be made to the product info saved in redux (simulates the backend).
-  The save as draft will call a function the is passed into the Product changes function. This will take the local state and pass it to the product array in the backend with status draft. (per my understanding). It will only be active if a change has been made.
- Similary, the submit button will only be active if a change has been made and will call a prop function. Depending on if the edit product prop is true or false it will either append to the products in redux (simulates the backend) or edit the adequate product.

6. Regarding treatment of button activity. Changes in product Info can easily be compared. 
```
  const [changes, setChanges] = useState(0)
  const [localData, setData] = useState(reduxData)

  const handleProductInfo = (key, val) => {
    if (localData[key] === reduxData[key] && reduxData[key] !== val) { // local data matches redux data but new value is different
      if (key === 'images') { 
        setChanges(prev => prev + 2) // this will keep image changes permanent
      }
      else {
        setChanges(prev => prev + 1)
      }
    }
    else if (localData[key] !== reduxData[key] && reduxData[key] === val) { // local data doesn't match reduc data but new value does
      if (key !== 'images') {
        setChanges(prev => prev + 1)
      }
    }
    setData((prevData) => ({
      ...prevData,
      [key]: val,
    }))
  }
```
This as above is just an example. Changes  > 0 may then be used to make disable the button or not.

------------------------------------------------------------------------------------
image
what is the use of this?

please check the way the checkout on main afroshop frontend is handled for clarity.
I'm available for any questions. Sorry this is late.

**Comment by ObiTam** [comment]
@Eshy10 please can we merge this and work on new changes in a new pr.

cc @EdiongAlexander @KaraEno please test

**Pull request merged**